tests/socket_test.c

```diff
@@ -804,17 +854,30 @@ static int s_test_outgoing_tcp_sock_error(struct aws_allocator *allocator, void
     struct aws_socket outgoing;
     ASSERT_SUCCESS(aws_socket_init(&outgoing, allocator, &options));
     /* tcp connect is non-blocking, it should return success, but the error callback will be invoked. */
-    ASSERT_SUCCESS(aws_socket_connect(&outgoing, &endpoint, event_loop, s_null_sock_connection, &args));
```

Comment:
I don't think so. If `aws_socket_connect` fails with an error, our contract should be that callback will not be triggered.

Reply:
im not necessarily talking about triggering a callback here. 
but if on freebsd we can get AWS_IO_SOCKET_CONNECTION_REFUSED right away instead of callback, should we document it and explicitly check here that last error was AWS_IO_SOCKET_CONNECTION_REFUSED?

Reply:
On looking more deeply, I think this is a problem with the [O_NONBLOCK](https://github.com/awslabs/aws-c-io/blob/main/source/posix/socket.c#L154) flag rather than the connect timeout. FreeBSD doesn't seem to respect it. I have reverted the change here for now and will try to create another PR after looking more into it, or maybe trying SOCK_NONBLOCK instead of O_NONBLOCK.

Reply:
On FreeBSD, the connect is failing sync with error  51/202 Test  #51: outgoing_tcp_sock_error 
```
***FAILURE*** [DEBUG] [2024-06-26T17:55:17Z] [00000b9aa7412000] [Unknown] - id=0x8209262b0 fd=7: connect failed with error code 61.
```. error code 61 is ECONNREFUSED.

Reply:
Created https://github.com/awslabs/aws-c-io/pull/648.